proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,649 @@
+
+* A numeric literal ``42`` is mapped to a promoted numeric literal and
+  requires the ``DataKinds`` extension.
+
+* A fractional numeric literal ``3.14`` cannot be mapped at the
+  moment, as we do not have promoted fractional numeric literals.
+
+* An unboxed numeric literal ``1337#`` cannot be mapped at the moment,
+  as we do not have promoted unboxed types.
+
+* A string literal ``"Hello"`` is mapped to a promoted string literal
+  ``"Hello"`` and requires the ``DataKinds`` extension.
+
+* A character literal ``'x'`` cannot be mapped at the moment, as we do
+  not have promoted character literals.
+
+* A tuple ``(a, b)`` is mapped to a promoted tuple ``'(a, b)``
+  and requires the ``DataKinds`` extension.
+
+* An unboxed tuple ``(# a, b #)`` cannot be mapped at the moment, as we
+  do not have promoted unboxed types.
+
+* A list literal ``[a, b, c]`` is mapped to a promoted list ``[a, b, c]``
+  and requires the ``DataKinds`` extension.
+
+* Function application ``f a`` is mapped to type-level function
+  application ``f a``.
+
+* Type application ``f @a`` is mapped to type-level type application
+  ``f @a`` and requires the ``TypeApplications`` extension.
+
+* Operator application ``x + y`` is mapped to type-level operator
+  application ``x + y`` and requires the ``TypeOperators`` extension.
+
+* A type signature ``a :: t`` is mapped to a kind signature ``a :: t``
+  and requires the ``KindSignatures`` extension.
+
+* Lambda functions ``\x -> b`` are not mapped and their use is an
+  error, as we do not have type-level lambdas at the moment.
+
+* Case-expressions ``case x of ...`` are not mapped and their use is
+  an error, as we do not have type-level case-expressions.
+
+* If-expressions ``if c then a else b`` are not mapped and their use
+  is an error, as we do not have type-level if-expressions.
+
+* In the same spirit, other syntactic constructs are mapped when
+  there's a direct type-level equivalent, and their use is an error
+  otherwise.
+
+Examples
+--------
+
+Compile-time color literals
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Definition site::
+
+  type family ParseRGB (s :: Symbol) :: (Nat, Nat, Nat) where
+    ...
+
+  type KnownRGB :: (Nat, Nat, Nat) -> Constraint
+  class KnownRGB c where
+    _rgbVal :: (Word8, Word8, Word8)
+
+  rgb :: forall s -> KnownRGB (ParseRGB c) => (Word8, Word8, Word8)
+  rgb s = _rgbVal @(ParseRGB s)
+
+Use site::
+
+  ghci> rgb "red"
+  (255, 0, 0)
+
+  ghci> rgb "#112233"
+  (17, 34, 51)
+
+  ghci> rgb "asdfasdf"
+  -- custom type error from ParseRGB
+
+Effect and Interactions
+-----------------------
+
+* Visible ``forall`` becomes available in types of terms, making them more similar to
+  types of types. There remains a discrepancy that ``forall`` for types is actually a
+  relevant quantifier, while the proposed ``forall x ->`` for terms is
+  irrelevant. This is to be resolved in the future by making type-level
+  ``forall`` irrelevant.
```

Comment:
It's a matter of reading the preposition "for". I say "`forall x ->` *for* types" rather than "`forall x ->` *in* types". The intended meaning is "`forall x ->` used to quantify over types", i.e. the kind-level quantifier.

Reply:
Sure. But are you positive that it is the most likely reading? Using consistent language throughout the proposal would ensure that there is no misunderstanding possible.

Reply:
I changed "In three releases" to "After three releases with the warning"

Reply:
I changed the paragraph to use "types of types" and "types of terms".